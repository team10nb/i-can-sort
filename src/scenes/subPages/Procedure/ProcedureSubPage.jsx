/*
    Author: Yuting Jiang, Yijie Lu

    The template of procedure subpages. Consists of a menu bar with the theme color and Card components to contain content.
*/

import React from 'react';
import PropTypes from 'prop-types';
import SwipeableViews from 'react-swipeable-views';
import { makeStyles, createMuiTheme, ThemeProvider } from '@material-ui/core/styles';
import AppBar from '@material-ui/core/AppBar';
import Tabs from '@material-ui/core/Tabs';
import Tab from '@material-ui/core/Tab';
import Typography from '@material-ui/core/Typography';
import Box from '@material-ui/core/Box';
import Card from '@material-ui/core/Card';
import { CardContent } from '@material-ui/core';
import HomeIcon from '@material-ui/icons/Home';
import Button from '@material-ui/core/Button';

//Set css
const useStyles = makeStyles((theme) => ({
  tabs:{
    textAlign: 'center',
    height: 60,
  },
  tab:{
    textAlign: 'center',
    height: 60,
    fontSize: 16,
  },
  icon:{
    textAlign: 'left',
    width: 100,
  },
  div:{
    display: 'flex',    
    width: 1020,
    '& > *': {
      margin: theme.spacing(3),
    }
  },
  cardOne:{
    marginTop: 80,
    background: "#F0F0F0",
    width: "35%",
    height: 520
  },
  cardTwo:{
    marginTop: 80,
    background: "#F0F0F0",
    width: "65%",
    height: 520,
    alignItems:"center",
  },
  cardOperation:{
    marginTop: 80,
    background: "#F0F0F0",
    width: "100%",
    height: 520,
    alignItems:"center",
  },

}));

//The subpages are corresponding to menu items
function TabPanel(props) {
  const { children, value, index, ...other } = props;

  return (
    <div
      role="tabpanel"
      hidden={value !== index}
      id={`full-width-tabpanel-${index}`}
      aria-labelledby={`full-width-tab-${index}`}
      {...other}
    >
      {value === index && (
        <Box p={3}>
          <Typography>{children}</Typography>
        </Box>
      )}
    </div>
  );
}

//Limit the prop type for TabPanel
TabPanel.propTypes = {
  children: PropTypes.object,
  index: PropTypes.any.isRequired,
  value: PropTypes.any.isRequired,
};

//Set props for tabs
function a11yProps(index) {
  return {
    id: `full-width-tab-${index}`,
    'aria-controls': `full-width-tabpanel-${index}`,
  };
}

//Return a menu bar consists of a home button and three tabs
//Each tab is corresponding to a tabpanel
export default function ProcedureSubPage(props) {
  const{color, algorithm, progress, history, intro, operate, Implementation} = props;

  const classes = useStyles();
  const theme = createMuiTheme({
    palette: {
      primary: {
        main: color,
      },
      secondary: {
        main: "#FFFFFF"
      },
      
    },
    
  });

  const [value, setValue] = React.useState(1);

  function finishPage(page){
    progress[page] = true;
    localStorage.setItem(algorithm, JSON.stringify(progress));
  }

  const handleClick = () => {
    //return to procedure main page 
    history.push({pathname: '/ProcedureMainPage'});
  };

  const handleChange = (event, newValue) => {
    //change the menu item, finish current page
    finishPage(newValue-1);
    setValue(newValue);
  };

  const handleChangeIndex = (index) => {
    //change the tabpanel
    setValue(index);
  };

  return (
   <div >
      <ThemeProvider theme={theme}>
      <AppBar>
        <Tabs

          value={value}
          onChange={handleChange}
          indicatorColor="secondary"
          background = "primary"
          
          variant="fullWidth"
          aria-label="full width tabs example"
        >
          <Button className={classes.icon} onClick={handleClick}>
            <HomeIcon style={{ fontSize: 30, color: "#FFFFFF"}} />
          </Button>

          <Tab label="Introduction" {...a11yProps(1)} className={classes.tab}/>
          <Tab label="Operation" {...a11yProps(2)} className={classes.tab}/>
          <Tab label="Implementation" {...a11yProps(3)} className={classes.tab}/>
        </Tabs>
      </AppBar>

      <SwipeableViews
        axis={theme.direction === 'rtl' ? 'x-reverse' : 'x'}
        index={value}
        onChangeIndex={handleChangeIndex}
      >
        <TabPanel value={value} index={0}> 
        </TabPanel>

        <TabPanel value={value} index={1} dir={theme.direction} {...finishPage(0)}>
          <div className = {classes.div}>
            <Card className={classes.cardOne}>
              <CardContent style={{paddingLeft: "24px", paddingTop: "0px", paddingRight:"23px"}}>
                {intro.introMessage}
              </CardContent>
            </Card>
            <Card className={classes.cardTwo} >
              <CardContent >
                {intro.animation}
              </CardContent>
            </Card>
          </div>
        </TabPanel>

        <TabPanel value={value} index={2} dir={theme.direction}>
          <div className = {classes.div}>
            <Card className={classes.cardOperation}>
              <CardContent>
                {operate}
              </CardContent>
            </Card>
          </div>
        </TabPanel>

        <TabPanel value={value} index={3} dir={theme.direction}>
            {Implementation}
        </TabPanel>
        
      </SwipeableViews>
      </ThemeProvider>
    </div>
  );
}
