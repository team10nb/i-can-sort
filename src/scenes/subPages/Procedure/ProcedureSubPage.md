
### **Developer**
Yuting Jiang, Yijie Lu

Reference: MATERIAL-UI (https://material-ui.com)


###  **Introduction**

This is a template for procedure subpages. Consists of three subpages, which are Introduction, Operation, and Implementation.

###  **Import**

```html
import SwipeableViews from 'react-swipeable-views';
import AppBar from '@material-ui/core/AppBar';
import Tabs from '@material-ui/core/Tabs';
import Tab from '@material-ui/core/Tab';
import Typography from '@material-ui/core/Typography';
import Box from '@material-ui/core/Box';
import Card from '@material-ui/core/Card';
import { CardContent } from '@material-ui/core';
import HomeIcon from '@material-ui/icons/Home';
import Button from '@material-ui/core/Button';
```

###  **Components**

| Name | Description | Child components |
| ---- | ----------- | ---------------- |
| AppBar | The bar contains three subpage's title      | HomeButton, Tab           | 
| Introduction Card | Contain introduction for a sorting algorithm        |                  | 
| Animation Card | Contain an example animation of a sorting algorithm        |                  | 

###  **Applications**

| Name | Description | 
| ---- | ----------- | 
| ProcedureBubble | The subpage for leaning Bubble sort |  
| ProcedureInsertion | The subpage for leaning Insertion sort |  
| ProcedureSelection | The subpage for leaning Selection sort |  
| ProcedureHeap | The subpage for leaning Heap sort |  
| ProcedureMerge | The subpage for leaning Merge sort |  
| ProcedureQuick | The subpage for leaning Quick sort |  


