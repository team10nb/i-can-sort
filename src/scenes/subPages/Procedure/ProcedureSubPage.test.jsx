
import { render } from "@testing-library/react";
import ProcedureSubPage from "./ProcedureSubPage";

const handleClick = jest.fn();
const color = '#1565c0';
const props = { handleClick, color};

test('should match snapshot', () => {

    const content = render(<ProcedureSubPage {...props}/>);
    expect(content).toMatchSnapshot();

})

test('should call  button',() => {
    
    const {getAllByRole} = render(<ProcedureSubPage {...props}/>);
    const buttons = getAllByRole('button');
    expect(buttons.length).toBe(1);

})

