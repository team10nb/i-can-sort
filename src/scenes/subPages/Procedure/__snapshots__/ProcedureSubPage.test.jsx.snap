// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <header
          class="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary MuiPaper-elevation4"
        >
          <div
            background="primary"
            class="MuiTabs-root"
          >
            <div
              class="MuiTabs-scroller MuiTabs-fixed"
              style="overflow: hidden;"
            >
              <div
                aria-label="full width tabs example"
                class="MuiTabs-flexContainer"
                role="tablist"
              >
                <button
                  class="MuiButtonBase-root MuiButton-root MuiButton-text makeStyles-icon-3 MuiButton-fullWidth"
                  tabindex="0"
                  textcolor="inherit"
                  type="button"
                  value="0"
                >
                  <span
                    class="MuiButton-label"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root"
                      focusable="false"
                      style="font-size: 30px; color: rgb(255, 255, 255);"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z"
                      />
                    </svg>
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </button>
                <button
                  aria-controls="full-width-tabpanel-1"
                  aria-selected="true"
                  class="MuiButtonBase-root MuiTab-root MuiTab-textColorInherit makeStyles-tab-2 Mui-selected MuiTab-fullWidth"
                  id="full-width-tab-1"
                  role="tab"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="MuiTab-wrapper"
                  >
                    Introduction
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </button>
                <button
                  aria-controls="full-width-tabpanel-2"
                  aria-selected="false"
                  class="MuiButtonBase-root MuiTab-root MuiTab-textColorInherit makeStyles-tab-2 MuiTab-fullWidth"
                  id="full-width-tab-2"
                  role="tab"
                  tabindex="-1"
                  type="button"
                >
                  <span
                    class="MuiTab-wrapper"
                  >
                    Operation
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </button>
                <button
                  aria-controls="full-width-tabpanel-3"
                  aria-selected="false"
                  class="MuiButtonBase-root MuiTab-root MuiTab-textColorInherit makeStyles-tab-2 MuiTab-fullWidth"
                  id="full-width-tab-3"
                  role="tab"
                  tabindex="-1"
                  type="button"
                >
                  <span
                    class="MuiTab-wrapper"
                  >
                    Implementation
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </button>
              </div>
              <span
                class="PrivateTabIndicator-root-7 PrivateTabIndicator-colorSecondary-9 MuiTabs-indicator"
                style="left: 0px; width: 0px;"
              />
            </div>
          </div>
        </header>
        <div
          style="overflow-x: hidden;"
        >
          <div
            class="react-swipeable-view-container"
            style="flex-direction: row; transition: all 0s ease 0s; direction: ltr; display: flex; will-change: transform;"
          >
            <div
              aria-hidden="false"
              data-swipeable="true"
              style="width: 100%; flex-shrink: 0; overflow: auto;"
            >
              <div
                aria-labelledby="full-width-tab-1"
                dir="ltr"
                id="full-width-tabpanel-1"
                role="tabpanel"
              >
                <div
                  class="MuiBox-root MuiBox-root-11"
                >
                  <p
                    class="MuiTypography-root MuiTypography-body1"
                  >
                    <div
                      class="makeStyles-div-4"
                    >
                      <div
                        class="MuiPaper-root MuiCard-root makeStyles-cardOne-5 MuiPaper-elevation1 MuiPaper-rounded"
                      >
                        <div
                          class="MuiCardContent-root"
                        />
                      </div>
                      <div
                        class="MuiPaper-root MuiCard-root makeStyles-cardTwo-6 MuiPaper-elevation1 MuiPaper-rounded"
                      />
                    </div>
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <header
        class="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary MuiPaper-elevation4"
      >
        <div
          background="primary"
          class="MuiTabs-root"
        >
          <div
            class="MuiTabs-scroller MuiTabs-fixed"
            style="overflow: hidden;"
          >
            <div
              aria-label="full width tabs example"
              class="MuiTabs-flexContainer"
              role="tablist"
            >
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-text makeStyles-icon-3 MuiButton-fullWidth"
                tabindex="0"
                textcolor="inherit"
                type="button"
                value="0"
              >
                <span
                  class="MuiButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    style="font-size: 30px; color: rgb(255, 255, 255);"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
              <button
                aria-controls="full-width-tabpanel-1"
                aria-selected="true"
                class="MuiButtonBase-root MuiTab-root MuiTab-textColorInherit makeStyles-tab-2 Mui-selected MuiTab-fullWidth"
                id="full-width-tab-1"
                role="tab"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiTab-wrapper"
                >
                  Introduction
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
              <button
                aria-controls="full-width-tabpanel-2"
                aria-selected="false"
                class="MuiButtonBase-root MuiTab-root MuiTab-textColorInherit makeStyles-tab-2 MuiTab-fullWidth"
                id="full-width-tab-2"
                role="tab"
                tabindex="-1"
                type="button"
              >
                <span
                  class="MuiTab-wrapper"
                >
                  Operation
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
              <button
                aria-controls="full-width-tabpanel-3"
                aria-selected="false"
                class="MuiButtonBase-root MuiTab-root MuiTab-textColorInherit makeStyles-tab-2 MuiTab-fullWidth"
                id="full-width-tab-3"
                role="tab"
                tabindex="-1"
                type="button"
              >
                <span
                  class="MuiTab-wrapper"
                >
                  Implementation
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
            <span
              class="PrivateTabIndicator-root-7 PrivateTabIndicator-colorSecondary-9 MuiTabs-indicator"
              style="left: 0px; width: 0px;"
            />
          </div>
        </div>
      </header>
      <div
        style="overflow-x: hidden;"
      >
        <div
          class="react-swipeable-view-container"
          style="flex-direction: row; transition: all 0s ease 0s; direction: ltr; display: flex; will-change: transform;"
        >
          <div
            aria-hidden="false"
            data-swipeable="true"
            style="width: 100%; flex-shrink: 0; overflow: auto;"
          >
            <div
              aria-labelledby="full-width-tab-1"
              dir="ltr"
              id="full-width-tabpanel-1"
              role="tabpanel"
            >
              <div
                class="MuiBox-root MuiBox-root-11"
              >
                <p
                  class="MuiTypography-root MuiTypography-body1"
                >
                  <div
                    class="makeStyles-div-4"
                  >
                    <div
                      class="MuiPaper-root MuiCard-root makeStyles-cardOne-5 MuiPaper-elevation1 MuiPaper-rounded"
                    >
                      <div
                        class="MuiCardContent-root"
                      />
                    </div>
                    <div
                      class="MuiPaper-root MuiCard-root makeStyles-cardTwo-6 MuiPaper-elevation1 MuiPaper-rounded"
                    />
                  </div>
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
